File: scripts/check-no-server.ts
import * as fs from "fs";
import * as path from "path";

// #EXT-003-REQ-3 Start
/**
 * No-server architecture check (structural enforcement of [REQ-3]).
 *
 * Agents are lightweight in-process threads, never per-agent services. This
 * checker scans `src/runtime/**` and any agent code and FAILS (exits non-zero)
 * if it finds a per-agent server/listener footprint, namely:
 *
 *   - `http.createServer` / `https.createServer` / `http2.create*Server`
 *   - `net.createServer` / `tls.createServer` / `dgram.createSocket`
 *   - a `.listen(` call (binding a socket/port)
 *   - `new WebSocketServer` / `new Server(` style socket bindings
 *
 * Scanned scope:
 *   - everything under `src/runtime/**`
 *   - any module whose path marks it as agent code (an `agent`/`agents` path
 *     segment, or a filename containing `agent`)
 *
 * Tests (`*.test.ts`) are excluded. If there is nothing to scan, the check
 * passes (exit 0).
 */
const ROOT = path.resolve(__dirname, "..", "..");
const SRC = path.join(ROOT, "src");
const RUNTIME_DIR = path.join(SRC, "runtime");

/** Source patterns that indicate a per-agent server/listener footprint. */
const SERVER_PATTERNS: ReadonlyArray<{ re: RegExp; reason: string }> = [
  { re: /\b(?:https?|http2)\.create(?:Secure)?Server\s*\(/, reason: "creates an HTTP server" },
  { re: /\bnet\.createServer\s*\(/, reason: "creates a TCP server" },
  { re: /\btls\.createServer\s*\(/, reason: "creates a TLS server" },
  { re: /\bdgram\.createSocket\s*\(/, reason: "creates a UDP socket" },
  { re: /\.listen\s*\(/, reason: "binds a listening socket/port via .listen(" },
  { re: /\bnew\s+WebSocketServer\b/, reason: "starts a WebSocket server" },
  { re: /\bnew\s+(?:WebSocket\.)?Server\s*\(/, reason: "starts a socket server" },
];

interface Violation {
  file: string;
  line: number;
  text: string;
  reason: string;
}

/** Recursively collect all `.ts` files under a directory (excluding tests). */
function collectTsFiles(dir: string): string[] {
  if (!fs.existsSync(dir)) {
    return [];
  }
  const out: string[] = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      out.push(...collectTsFiles(full));
    } else if (
      entry.isFile() &&
      entry.name.endsWith(".ts") &&
      !entry.name.endsWith(".test.ts")
    ) {
      out.push(full);
    }
  }
  return out;
}

/** True if a file (by its path) counts as agent code we must police. */
function isAgentCode(file: string): boolean {
  const rel = path.relative(SRC, file).replace(/\\/g, "/");
  const segments = rel.split("/");
  const dirSegments = segments.slice(0, -1);
  const base = segments[segments.length - 1];
  if (dirSegments.some((s) => s === "agent" || s === "agents")) {
    return true;
  }
  return /agent/i.test(base);
}

function main(): void {
  const runtimeFiles = collectTsFiles(RUNTIME_DIR);
  const allSrcFiles = collectTsFiles(SRC);
  const agentFiles = allSrcFiles.filter(isAgentCode);

  const fileSet = new Set<string>([...runtimeFiles, ...agentFiles]);
  const files = [...fileSet];

  const violations: Violation[] = [];

  for (const file of files) {
    const lines = fs.readFileSync(file, "utf8").split(/\r?\n/);
    const relFile = path.relative(ROOT, file).replace(/\\/g, "/");

    lines.forEach((text, i) => {
      for (const { re, reason } of SERVER_PATTERNS) {
        if (re.test(text)) {
          violations.push({ file: relFile, line: i + 1, text: text.trim(), reason });
          break;
        }
      }
    });
  }

  if (violations.length > 0) {
    console.error(
      "No-server violation: agents are lightweight threads and must not start a server/listener."
    );
    for (const v of violations) {
      console.error(`  ${v.file}:${v.line}  ${v.reason}\n    ${v.text}`);
    }
    process.exit(1);
  }

  console.log(
    `check-no-server: OK — scanned ${files.length} runtime/agent file(s); no per-agent server/listener footprint.`
  );
  process.exit(0);
}

main();
// #EXT-003-REQ-3 End

File: scripts/check_no_server.py
"""Structural no-server architecture check (EXT-003 / REQ-3).

Agents are lightweight threads, never per-agent services. This check scans the
runtime (``jaros/runtime/**.py``) and any agent code (``jaros/agents/**.py``)
and fails (non-zero exit) if any of them opens a listening socket or HTTP
server. Detected patterns:

- ``socket.socket`` (constructing a raw socket);
- a ``.bind(`` or ``.listen(`` call (binding/listening a socket);
- ``HTTPServer`` (``http.server.HTTPServer`` / ``ThreadingHTTPServer``);
- the ``socketserver`` module;
- ``asyncio.start_server`` (an asyncio listening server).

Exits 0 on a clean tree and when there are no scannable files. Pattern matching
is done on the AST (imports + call expressions), so comments and strings do not
trip it.

Run as: ``python scripts/check_no_server.py``
"""

from __future__ import annotations

import ast
import sys
from pathlib import Path

# #EXT-003-REQ-3 Start
# Directories whose Python files must be free of server/listener footprints.
SCAN_DIRS = (
    Path("jaros") / "runtime",
    Path("jaros") / "agents",
)

# Module imports that imply a per-agent server footprint.
FORBIDDEN_IMPORT_PREFIXES = ("socketserver",)

# Dotted attribute calls that imply opening a server/listener, e.g.
# ``socket.socket(...)`` or ``asyncio.start_server(...)``.
FORBIDDEN_DOTTED_CALLS = {
    ("socket", "socket"),
    ("asyncio", "start_server"),
}

# Bare names that imply an HTTP server when called or imported, e.g.
# ``HTTPServer(...)`` / ``ThreadingHTTPServer(...)``.
FORBIDDEN_SERVER_NAMES = {"HTTPServer", "ThreadingHTTPServer"}

# Method names that, when called on anything, imply binding/listening a socket.
FORBIDDEN_METHODS = {"bind", "listen"}


def _dotted(node: ast.expr) -> tuple[str, ...]:
    """Return the dotted-name tuple of an attribute/name chain, else ()."""
    parts: list[str] = []
    cur: ast.expr | None = node
    while isinstance(cur, ast.Attribute):
        parts.append(cur.attr)
        cur = cur.value
    if isinstance(cur, ast.Name):
        parts.append(cur.id)
        parts.reverse()
        return tuple(parts)
    return ()


def _check_call(node: ast.Call, path: Path) -> list[str]:
    violations: list[str] = []
    func = node.func
    if isinstance(func, ast.Attribute):
        if func.attr in FORBIDDEN_METHODS:
            violations.append(
                f"{path.as_posix()}:{node.lineno}: forbidden socket "
                f"'.{func.attr}(' call"
            )
        dotted = _dotted(func)
        if len(dotted) >= 2 and (dotted[-2], dotted[-1]) in FORBIDDEN_DOTTED_CALLS:
            violations.append(
                f"{path.as_posix()}:{node.lineno}: forbidden server call "
                f"{'.'.join(dotted[-2:])!r}"
            )
    elif isinstance(func, ast.Name) and func.id in FORBIDDEN_SERVER_NAMES:
        violations.append(
            f"{path.as_posix()}:{node.lineno}: forbidden server "
            f"construction {func.id!r}"
        )
    return violations


def _check_import(name: str | None, path: Path, lineno: int) -> list[str]:
    if not name:
        return []
    if any(name == p or name.startswith(p + ".") for p in FORBIDDEN_IMPORT_PREFIXES):
        return [f"{path.as_posix()}:{lineno}: forbidden server import {name!r}"]
    return []


def find_violations(repo_root: Path) -> list[str]:
    """Return human-readable server-footprint violations (empty if clean)."""
    violations: list[str] = []
    for rel_dir in SCAN_DIRS:
        scan_dir = repo_root / rel_dir
        if not scan_dir.is_dir():
            continue
        for path in sorted(scan_dir.rglob("*.py")):
            source = path.read_text(encoding="utf-8")
            tree = ast.parse(source, filename=str(path))
            for node in ast.walk(tree):
                if isinstance(node, ast.Call):
                    violations.extend(_check_call(node, path))
                elif isinstance(node, ast.Import):
                    for alias in node.names:
                        violations.extend(
                            _check_import(alias.name, path, node.lineno)
                        )
                elif isinstance(node, ast.ImportFrom):
                    if node.level == 0:
                        violations.extend(
                            _check_import(node.module, path, node.lineno)
                        )
                    for alias in node.names:
                        if alias.name in FORBIDDEN_SERVER_NAMES:
                            violations.append(
                                f"{path.as_posix()}:{node.lineno}: forbidden "
                                f"server import {alias.name!r}"
                            )
    return violations


def main() -> int:
    repo_root = Path(__file__).resolve().parent.parent
    violations = find_violations(repo_root)
    if violations:
        print("No-server check FAILED:", file=sys.stderr)
        for v in violations:
            print(f"  {v}", file=sys.stderr)
        return 1
    print("No-server check passed: no per-agent server/listener footprint.")
    return 0
# #EXT-003-REQ-3 End


if __name__ == "__main__":
    sys.exit(main())
